### ISSUE TITLE
[Question]: "Authentication" with multiple projects/browser in config randomly fails

### ISSUE THREAD

**Issue by radekBednarik** [microsoft/playwright#8064]
### Your question

Hello, I have a rather lengthy question, so please bear with me :)

I am working on generalized set of tests which will run on multiple websites - these sites are each on different domain and can have slightly modified content. 

To speed things up, I want on each site to interact with cookie banner only once - that means to allow cookie consents only once, save the context state and then use it for all tests on that domain, so I do not have to lose time with confirming the consents every time the fresh context for each test loads.

Now since my tests are heavily parametrized, I cannot use the process described here https://playwright.dev/docs/auth#reuse-authentication-in-playwright-test

What I did was this:

1. Create global setup, which will create "state.json" file with empty object {}.
2. Create global teardown, which will delete the "state.json" file after all is done.
3. Override `page` fixture like this to save the storage state, only if the .json file is empty, which should be only at the very beginning before the first test:

```
module.exports = test.extend({
  url: undefined,
  page: async ({ page, url }, use) => {
    if (isContextStateEmpty()) {
      const entryPage = new BasePage(page, { cookieManager: CookieManager });
      await entryPage.goto(url);
      await entryPage.cookieManager.clickBttnMainBannerAccept();
      await entryPage.wait();
      await entryPage.context.storageState({ path: "state.json" });
    }
    await use(page);
  },
});
```
4. Now I can create test specific fixtures, which will extend the above fixture, e.g.:

```
module.exports = test.extend({
  url: undefined,
  container: undefined,
  gaProperty: undefined,
  storageState: "state.json",
  testedPage: async ({ page, container, url, gaProperty }, use) => {
    const testedPage = new BasePage(page, { cookieManager: CookieManager }, { network: Network });

    await Promise.all([
      testedPage.network.catchRequest(re_pageview, container, undefined, undefined, gaProperty),
      testedPage.goto(url),
    ]);
    await use(testedPage);
    container.length = 0;
  },
});
```
5. this fixture is then used by the test spec files.

ISSUE:

If I have in my `playwright.config.js` only one project, e.g. `chromium`, or `firefox` or `webkit`, this is working as expected. The page with consent banner will appear, consents are accepted, and then next pages are visited and banner is not displayed.

However, if I will have at least two different browser specified as `projects` in the config file, this implementation will sometimes run without issue, sometimes it will stop working after visiting several pages of the site - that means, that Coookie Consent banner will start appearing again.

I do not know, where to look for potential issue. Any idea or pointers would be greatly appreciated.

**Comment by pavelfeldman** [comment]
Looks like you have some timing issue with the cookie handling that triggers races. We can only help if there is a reduction we can reproduce.

**Comment by radekBednarik** [comment]
@pavelfeldman i will try to prepare a small repro repo, pls gimme bit of time

**Comment by radekBednarik** [comment]
@pavelfeldman hello, here it is https://github.com/radekBednarik/pl-issue-repro if anything else is needed, please let me know. Thank you for the help, much appreciated

**Comment by pavelfeldman** [comment]
Sorry about the delay. There is a lot happening in this test and it feels like the same could be achieved with less fixture overrides. For example, you could auth in the global setup like so: https://playwright.dev/docs/next/auth#reuse-authentication-in-playwright-test. Could you give it a try? You wound't need any fixture overrides for that.

**Comment by radekBednarik** [comment]
@pavelfeldman hi, no problem. I will try hopefully today/tomorrow and let you know. Thanks

**Comment by radekBednarik** [comment]
@pavelfeldman hello, the proposed auth reuse is known to me, I mentioned it in my first comment. The main drawback of this approach for my use case is, that (unfortunately) I am working with web, which is implemented across large number of domains. Each domain is different legal entity and consent accepted on one will not work on the other ones. It would be very bad, if this would actually happen ;)

I could use auth reuse, as described in the docs, but that would mean for me to add another parametrization probably via ENV variables and more complexity in the CI pipelines, because I have to pass the initial url to the globalSetup.

That is why I tried the approach you saw in the repo reduction. 

I assumed, that if my approach works in the case of single browser type and in case of multiple browser types running in parallel there are issues, then there may be some problem e.g. in the test runner? But this is beyond my current knowledge to assess.

**Comment by pavelfeldman** [comment]
I re-read your request. I think the issue here is that in case of multiple projects, these tests run in parallel. As a result, on all three workers, first test will see no state and will try logging in. You should make the name of the file unique, you can use browserName fixture + file name combination for that for example.

**Comment by pavelfeldman** [comment]
Closing this one, but please feel free to open another issue if there are more problems!
